sympy/functions/elementary/tests/test_piecewise.py

```diff
@@ -1226,3 +1226,54 @@ def test_issue_8458():
     # Test for problem highlighted during review
     p3 = Piecewise((x+1, Eq(x, -1)), (4*x + (y-2)**4, Eq(x, 0) & Eq(x+y, 2)), (sin(x), True))
     assert p3.simplify() == Piecewise((0, Eq(x, -1)), (sin(x), True))
+
+
+def test_eval_rewrite_as_KroneckerDelta():
+    x, y, z, n, t = symbols('x y z n t')
+    K = KroneckerDelta
+    f = lambda p: p.rewrite(K)
+
+    p1 = Piecewise((0, Eq(x, y)), (1, True))
+    assert f(p1).equals(1 - K(x, y))
```

Comment:
Are you using `.equals` because `==` won't work? Generally `==` is much more efficient and gives a more reliable test of the output since it only matches the exact form.

I think it would be cleaner just to write
```
    assert p1.rewrite(K) == 1 - K(x, y)
```

Reply:
I was running into this issue:

```
>>> from sympy import *
>>> from sympy.abc import x, y
>>> p6 = Piecewise((0, Ne(x, 1) & Ne(y, 4)), (1, True))
>>> p6.rewrite(KroneckerDelta)
-(1 - KroneckerDelta(1, x))*(1 - KroneckerDelta(4, y)) + 1
>>> p6.rewrite(KroneckerDelta) == -(1 - KroneckerDelta(1, x))*(1 - KroneckerDelta(4, y)) + 1
False
>>> -(1 - KroneckerDelta(1, x))*(1 - KroneckerDelta(4, y)) + 1
(1 - KroneckerDelta(4, y))*(KroneckerDelta(1, x) - 1) + 1
>>> p6.rewrite(KroneckerDelta) == (1 - KroneckerDelta(4, y))*(KroneckerDelta(1, x) - 1) + 1
False
>>>
```

Reply:
I guess I could just expand both sides.

Reply:
This seems like a bug somehow but yes an expand could be used here.